storage/scratchpad/src/accumulator/mod.rs

```diff
@@ -123,6 +124,133 @@ where
+    ///                                               /          \
+    ///                                       old_root            siblings[2]
+    ///                                      /        \
+    ///                                    /            \
+    ///                                  /                \
+    ///                                /                    \
+    ///                              /                        \
+    ///                            /                            \
+    ///                          /                                \
+    ///                        /                                    \
+    /// frozen_subtree_roots[0]                                      o
+    ///                    /   \                                    / \
+    ///                   /     \                                  /   \
+    ///                  o       o                                o     siblings[1]
+    ///                 / \     / \                              / \
+    ///                o   o   o   o      frozen_subtree_roots[1]   siblings[0]
+    ///               / \ / \ / \ / \                         / \           / \
+    ///               o o o o o o o o                         o o           A B
+    /// ```
+    pub fn append_siblings(&self, siblings: &[HashValue]) -> Result<HashValue> {
+        Self::append_siblings_impl(&self.frozen_subtree_roots, self.num_leaves, siblings)
+    }
+
+    fn append_siblings_impl(
+        frozen_subtree_roots: &[HashValue],
+        num_existing_leaves: u64,
+        siblings: &[HashValue],
+    ) -> Result<HashValue> {
+        if frozen_subtree_roots.is_empty() {
+            ensure!(
+                siblings.len() == 1,
+                "{} siblings are provided when only one is required.",
+                siblings.len(),
+            );
+            return Ok(siblings[0]);
+        }
+        if siblings.is_empty() {
+            ensure!(
+                frozen_subtree_roots.len() == 1,
+                "No siblings are provided when some are required to compute root hash."
+            );
+            return Ok(frozen_subtree_roots[0]);
+        }
+
+        Self::validate_siblings(num_existing_leaves, siblings)?;
+
+        let mut frozen_subtree_iter = frozen_subtree_roots.iter().rev().peekable();
+        let mut sibling_iter = siblings.iter().peekable();
+        let mut current_hash = *sibling_iter
+            .next()
+            .expect("Code would have returned if siblings is empty.");
+
+        // Using the above example, num_existing_leaves = 10 = 0b1010. We check each bit from right
+        // to left starting from the rightmost one bit. If a bit is 0, it means there is no
+        // existing frozen subtree on this level, so we combine current hash with a sibling.
+        // Otherwise we combine it with the corresponding frozen subtree.
+        let mut bitmap = num_existing_leaves >> num_existing_leaves.trailing_zeros();
```

Comment:
add
`assert!(bitmap.ones() == frozen_subtree_roots.len());`?

Reply:
This is already checked in `new`.